src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs

```diff
@@ -307,7 +308,23 @@ public Task<Compilation> GetCompilationAsync(Solution solution, CancellationToke
                 }
                 else
                 {
-                    return GetOrBuildCompilationAsync(solution, lockGate: true, cancellationToken: cancellationToken);
+                    return GetOrBuildCompilationInfoAsync(solution, lockGate: true, cancellationToken: cancellationToken)
+                        .ContinueWith(t => t.Result.Compilation, cancellationToken, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+                }
+            }
+
+            public Task<bool> GetCompilationHasIncompleteReferencesAsync(Solution solution, CancellationToken cancellationToken)
+            {
+                var state = this.ReadState();
+
+                if (state.HasIncompleteReferences.HasValue)
```

Comment:
Where do I initialize the state with 'true'? InProgress state will return null, thus forcing the compilation to be finalized. The only override (today) that could return something other than null is in FinalState.

The intent of the API is for this to be as side-effect free as GetCompilation is, and calling them in either order shouldn't cause any effects. I _believe_ I got this right, but I confess that it took me a half of a day to understand how the code worked...